File: Top-Interview-Questions/Trees-and-Graphs-Medium-Collection/116-populating-next-right-pointers-in-each-node.js
/**
 * // Definition for a Node.
 * function Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {Node} root
 * @return {Node}
 * Runtime: 92 ms, faster than 76.89% of JavaScript online submissions for Populating Next Right Pointers in Each Node.
 * Memory Usage: 43.4 MB, less than 16.90% of JavaScript online submissions for Populating Next Right Pointers in Each Node.
 * TC:O(N), SC:O(logN)
 */
var connect = function(root) {
    if (!root || !root.left) {
        return root;
    }

    root.left.next = root.right;
    root.right.next = root.next ? root.next.left : null;

    connect(root.left);
    connect(root.right);

    return root;
};

File: Top-Interview-Questions/Trees-and-Graphs-Medium-Collection/116-Populating-Next-Right-Pointers-in-Each-Node.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

// Solution #1 BFS
// TC: O(n)
// SC: O(2^(height-1))
// Runtime: 48 ms, faster than 15.44% 
// Memory Usage: 17.2 MB, less than 27.04%
class Solution {
public:
    Node* connect(Node* root) {
        
        if(!root)
            return root;
        
        queue<Node*> q;
        q.push(root);
        int length;
        
        while(!q.empty()){
            
            length = q.size();
            
            for(int i = 0; i < length; i++){
                Node *node = q.front();
                q.pop();
                
                if(i < length-1)
                    node->next = q.front();
                
                if(node->left){
                    q.push(node->left);
                    q.push(node->right);
                }
            }
        }
        
        return root;
    }
};